### ISSUE TITLE
Lua bindings

### ISSUE THREAD

**Issue by tomsharratt** [skypjack/entt#21]
A common requirement for game engines is scripting (with lua in my case). As this ECS (and most other) use templates you can't bind most of the important functionality into lua.

For example, how would you bind `registry.get<T>(entity);` into lua?

**Comment by skypjack** [comment]
You can inherit from the registry or create a wrapper and export from and towards Lua the sole components you want to expose.
Probably the wrapper will have explicit methods like `get(CompA)`, `get(CompN)` and do on.
Lua and C++ are too much different languages to pretend to use an high performance tool directly.

**Comment by tomsharratt** [comment]
Might not be understanding you correctly.

Are you suggesting for each component I want to expose to Lua to create a wrapper. So for a Transform component I would create `getTransformComponent()`, `assignTransformComponent(...)` and so on? I know Lua can't work with templates, that is essentially the whole problem I am facing - how to wrap/abstract the ECS in a way which ideally doesn't require a ton of boilerplate code to be written for each component.

Thanks

**Comment by skypjack** [comment]
Well, think about this.  
If I hadn't used template, I had to write N member functions to assign components and N member functions to get them.  
Fortunately, C++ supports templates and I avoided writing all that code. Lua doesn't support template.  
You got it.

However, I wouldn't expose up to the transform through the scripting system.  
I mean, try to reduce to a minimum the components your script uses or define only one dedicated component for it. Then a system can write back and forth data from that component to all the others to which it maps. This will force you to write only a couple of function in a wrapper and a few lines in a system to keep data synced.
